macros/src/lib.rs

```diff
@@ -2,8 +2,234 @@ extern crate proc_macro;
 
 use proc_macro::TokenStream;
 use proc_macro2::Span;
-use quote::quote;
-use syn::{self, Data, Fields, GenericParam, LitInt, TypeParamBound};
+use quote::{format_ident, quote};
+use syn::{
+    self, parse_macro_input, Data, DeriveInput, Fields, FieldsNamed, FieldsUnnamed, GenericParam,
+    LitInt, TypeParamBound,
+};
+
+#[proc_macro_derive(ViewEnum)]
+pub fn derive_viewenum(input: TokenStream) -> TokenStream {
+    let ast = parse_macro_input!(input as DeriveInput);
+
+    let enum_name = &ast.ident;
+
+    let enum_data = if let Data::Enum(data) = &ast.data {
+        data
+    } else {
+        panic!("derive(ViewEnum) only supports enums");
+    };
+
+    let mut enum_field_counts = Vec::new();
+    let mut enum_translates = Vec::new();
+    let mut enum_indexes = Vec::new();
+    let mut enum_mut_indexes = Vec::new();
+    let mut enum_draws = Vec::new();
+
+    enum_data.variants.iter().for_each(|variant| {
+        let variant_name = &variant.ident;
+
+        let (enum_field_count, enum_translate, enum_index, enum_mut_index, enum_draw) =
+            match &variant.fields {
+                Fields::Named(FieldsNamed { named, .. }) => {
+                    let field_idents = named
+                        .iter()
+                        .map(|field| field.ident.as_ref().unwrap())
+                        .collect::<Vec<_>>();
+
+                    let fields_count = named.iter().count();
+                    let enum_field_count = quote! {
+                        Self::#variant_name { #(#field_idents,)* } => {
+                            #fields_count
+                        }
+                    };
+
+                    let translate_fields =
+                        field_idents.iter().map(|f| quote!(#f: #f.translate(by)));
+                    let enum_translate = quote! {
+                        Self::#variant_name { #(#field_idents,)* } => {
+                            Self::#variant_name {
+                                #(#translate_fields,)*
+                            }
+                        }
+                    };
+
+                    let fields_index = field_idents
+                        .iter()
+                        .enumerate()
+                        .map(|(i, f)| quote!(#i => #f,));
+                    let enum_index = quote! {
+                        Self::#variant_name { #(#field_idents,)* } => {
+                            match index {
+                                #(#fields_index)*
+                                _ => panic!("Invalid index!"),
```

Comment:
I would suggest to pick a panic message that is a bit more informative. Also, this message should be:
 - generated as an `#[inline(never)]` function to avoid including the string multiple times
 - it should be the same as the message for `Unnamed` variants

Reply:
Done

Reply:
@StefanBossbaly much better, thanks! This comment and the one about unit enums still stand, though. I can do these changes if you want me to, let me know if you do :)

Reply:
@bugadani This comment should be addressed. Both `at()` and `at_mut()` are now `#[inline(never)]` and return the same message (`Invalid index! Index exceeds the value returned by len()`)

Reply:
Hmm I don't think I communicated very well. But never mind, I don't want to force you to a lengthy back-and-forth on this.